### PR TITLE
Expire cache on deletion of case properties on Data Dictionary

### PR DESCRIPTION
**Pull request by mkangia**
## Product Description

For projects that are on advance plan or higher, When a user adds, updates or deletes a case property on Data dictionary, show them updated case properties on CLE.

## Technical Summary

https://dimagi.atlassian.net/browse/SC-3336

CLE has a list of column options which it populates by using the case properties for the project fetched by [all_case_properties_by_domain](https://github.com/dimagi/commcare-hq/blob/2408dd86005d11f70ecd795c7fdbc61255a74c23/corehq/apps/app_manager/app_schemas/case_properties.py#L512) which is cached for 5 minutes.
The change here expires that cache when case property is saved or deleted. The same expiration is done for Application changes, so extracted that as a function and called that.
A more ideal change would be to expire cache only for the case type of the case property but the way caching is set that isn't feasible.
Though since CLE only fetches case properties from Data Dictionary for projects that have the privilege for DD, the expiration was done only for such projects which are `Advanced Plan and higher` plans. So, the impact of the cache expiry would be low.

## Feature Flag

`DATA_DICTIONARY`

### Safety story

No functional change. Tested locally that CLE UI continues to work well.
The only concern was that expiring cache more frequently should not result in any performance issues on production which should not happen considering its only for advance plan and higher and additionally the updates to DD are rare.

### QA Plan

https://dimagi.atlassian.net/browse/QA-6148

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by mkangia** [comment]
please hold off on review

**Comment by mkangia** [comment]
All set for review now.

File: corehq/apps/app_manager/app_schemas/case_properties.py [hunk 1]
```diff
@@ -543,3 +543,10 @@ def _get_usercase_default_properties(domain):
 
     fields_def = CustomDataFieldsDefinition.get_or_create(domain, CUSTOM_USER_DATA_FIELD_TYPE)
     return [f.slug for f in fields_def.get_fields()]
+
+
+def expire_case_properties_caches(domain):
```
Comment (kaapstorm) [inline corehq/apps/app_manager/app_schemas/case_properties.py:548]:
```suggestion
def expire_case_properties_caches(domain):
    """
    Clears ``all_case_properties_by_domain()`` quickcache for ``domain``
    and all values of params ``include_parent_properties`` and ``exclude_deprecated_properties``.
    """
```

**Review by kaapstorm — DISMISSED** [review]
:+1: 

Non-blocking suggestion for a docstring to explain what those boolean values are about.

File: corehq/apps/app_manager/app_schemas/case_properties.py (hunk 1 above)
Comment (mkangia) [reply corehq/apps/app_manager/app_schemas/case_properties.py:548]: Thanks for the suggestion @kaapstorm 

That detail about passing all combinations of params is something to do with how quickcache works, which I agree isn't obvious, so adding a comment sounds right to me though I am not sure it should in the function documentation.

I would image documentation should be something like "Expire any cache related to case properties"?
Even though, for now, it is limited to one function, It's a good placeholder to add any other such caches and expire them all together at one place.

What do you think?

**Review by zandre-eng — DISMISSED** [review]
Looks good!

File: corehq/apps/app_manager/app_schemas/case_properties.py (hunk 1 above)
Comment (kaapstorm) [reply corehq/apps/app_manager/app_schemas/case_properties.py:548]: Yeah, a comment is fine. Using keyword arguments would also be fine.

I'm not sure about the next developer, but speaking for myself, I understood that we are clearing the cache for all combinations of the function's parameters. I just didn't know what those parameters were, and therefore why they needed to be cleared, without going to look them up.

**Review by kaapstorm — APPROVED** [review]
> kaapstorm previously approved these changes 5 days ago

I'm not sure why GitHub isn't showing this PR as approved by me any more. I still approve this PR. :)

**Comment by mkangia** [comment]
Hey @kaapstorm 

> I'm not sure why GitHub isn't showing this PR as approved by me any more. I still approve this PR. :)

That happened because the base branch was changed which marked the review as stale

File: corehq/apps/app_manager/app_schemas/case_properties.py (hunk 1 above)
Comment (mkangia) [reply corehq/apps/app_manager/app_schemas/case_properties.py:548]: Makes sense about kwargs.

https://github.com/dimagi/commcare-hq/pull/34154/commits/055a014e090dde959c91a47ea13ab7704fba9f1f

**Comment by mkangia** [comment]
Hey @kaapstorm 

I addressed your feedback on https://github.com/dimagi/commcare-hq/pull/34154#discussion_r1502326651.
Additionally there was a weird test failure that I fixed after your last review.
Please take another pass when you can. QA is done on this so this is all set to be released.

**Review by kaapstorm — APPROVED** [review]
Nice. Thank you.

**Pull request merged**